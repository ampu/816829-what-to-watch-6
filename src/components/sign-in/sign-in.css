.sign-in__form--disabled {
  filter: grayscale(100%);
}

.sign-in :disabled {
  cursor: default;
}
