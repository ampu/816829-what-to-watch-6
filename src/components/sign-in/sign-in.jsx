import React, {useState, useCallback} from 'react';
import PropTypes from 'prop-types';
import {Redirect} from 'react-router-dom';
import {connect} from 'react-redux';
import getClassName from 'classnames';

import Logo from '../logo/logo';
import {OperationStatus, OPERATION_STATUSES} from '../../constants/operation-status';
import {MainPath} from '../../constants/paths';
import {postLogin} from '../../store/operations/user-operations';

import './sign-in.css';

const Field = {
  EMAIL: `email`,
  PASSWORD: `password`,
};

const getFieldClassName = (field, error) => {
  return getClassName({
    [`sign-in__field`]: true,
    [`sign-in__field--error`]: error && error.field === field,
  });
};

const SignIn = ({onSubmit, loginStatus} = {}) => {

  const [error, setError] = useState(null);

  const [email, setEmail] = useState(``);
  const handleLoginChange = useCallback((evt) => {
    setEmail(evt.target.value);
  }, [setEmail]);

  const [password, setPassword] = useState(``);
  const handlePasswordChange = useCallback((evt) => {
    setPassword(evt.target.value);
  }, [setPassword]);

  const handleFormSubmit = useCallback((evt) => {
    evt.preventDefault();

    setError(null);

    onSubmit({email, password})
      .catch(({response = {}}) => {
        const {data = {}} = response;
        setError({message: data.error});
      });
  }, [onSubmit, email, password, setError]);

  const isDisabled = loginStatus === OperationStatus.PENDING;

  const formClassMap = {
    [`sign-in__form`]: true,
    [`sign-in__form--disabled`]: isDisabled,
  };

  if (loginStatus === OperationStatus.RESOLVED) {
    return <Redirect to={MainPath.INDEX}/>;
  }

  return (
    <div className="user-page">
      <header className="page-header user-page__head">
        <Logo/>

        <h1 className="page-title user-page__title">Sign in</h1>
      </header>

      <div className="sign-in user-page__content">
        <form action="#" className={getClassName(formClassMap)} onSubmit={handleFormSubmit}>
          {error && error.message && (
            <div className="sign-in__message">
              <p>{error.message}</p>
            </div>
          )}
          <div className="sign-in__fields">
            <div className={getFieldClassName(Field.EMAIL, error)}>
              <input className="sign-in__input" id="user-email"
                type="email" placeholder="Email address" name="user-email" autoComplete="username" required
                value={email} onChange={handleLoginChange} disabled={isDisabled}/>

              <label className="sign-in__label visually-hidden" htmlFor="user-email">Email address</label>
            </div>
            <div className={getFieldClassName(Field.PASSWORD, error)}>
              <input className="sign-in__input" id="user-password"
                type="password" placeholder="Password" name="user-password" autoComplete="current-password" required
                value={password} onChange={handlePasswordChange} disabled={isDisabled}/>

              <label className="sign-in__label visually-hidden" htmlFor="user-password">Password</label>
            </div>
          </div>
          <div className="sign-in__submit">
            <button className="sign-in__btn" type="submit" disabled={isDisabled}>Sign in</button>
          </div>
        </form>
      </div>

      <footer className="page-footer">
        <Logo isLight/>

        <div className="copyright">
          <p>Â© 2019 What to watch Ltd.</p>
        </div>
      </footer>
    </div>
  );
};

SignIn.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  loginStatus: PropTypes.oneOf(OPERATION_STATUSES),
  error: PropTypes.shape({
    message: PropTypes.string,
    field: PropTypes.oneOf(Object.values(Field)),
  }),
};

const mapStateToProps = (state) => ({
  loginStatus: state.loginStatus,
});

const mapDispatchToProps = (dispatch) => ({
  onSubmit({email, password}) {
    return dispatch(postLogin({email, password}));
  },
});

export {SignIn};
export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
