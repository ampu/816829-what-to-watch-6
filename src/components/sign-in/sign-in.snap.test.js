import React from 'react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';

import {render} from '@testing-library/react';

import {SignIn} from './sign-in';
import OperationStatus from '../../constants/operation-status';

describe(`SignIn component`, () => {

  const onSubmit = () => Promise.resolve();

  it(`Should render according to the snapshot while authorization status is unset or rejected`, () => {
    [
      undefined,
      OperationStatus.UNSET,
      OperationStatus.REJECTED,
    ].forEach((loginStatus) => {
      const {container} = render((
        <Router history={createMemoryHistory()}>
          <SignIn onSubmit={onSubmit} loginStatus={loginStatus}/>
        </Router>
      ));
      expect(container).toMatchSnapshot();
    });
  });

  it(`Should render according to the snapshot while authorization status is pending`, () => {
    const {container} = render((
      <Router history={createMemoryHistory()}>
        <SignIn onSubmit={onSubmit} loginStatus={OperationStatus.PENDING}/>
      </Router>
    ));
    expect(container).toMatchSnapshot();
  });
});
