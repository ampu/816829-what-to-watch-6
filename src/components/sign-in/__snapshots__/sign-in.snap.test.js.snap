// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn component Should render according to the snapshot while authorization status is pending 1`] = `
<div>
  <div
    class="user-page"
  >
    <header
      class="page-header user-page__head"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <h1
        class="page-title user-page__title"
      >
        Sign in
      </h1>
    </header>
    <div
      class="sign-in user-page__content"
    >
      <form
        action="#"
        class="sign-in__form sign-in__form--disabled"
      >
        <div
          class="sign-in__fields"
        >
          <div
            class="sign-in__field"
          >
            <input
              autocomplete="username"
              class="sign-in__input"
              disabled=""
              id="user-email"
              name="user-email"
              placeholder="Email address"
              required=""
              type="email"
              value=""
            />
            <label
              class="sign-in__label visually-hidden"
              for="user-email"
            >
              Email address
            </label>
          </div>
          <div
            class="sign-in__field"
          >
            <input
              autocomplete="current-password"
              class="sign-in__input"
              disabled=""
              id="user-password"
              name="user-password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
            <label
              class="sign-in__label visually-hidden"
              for="user-password"
            >
              Password
            </label>
          </div>
        </div>
        <div
          class="sign-in__submit"
        >
          <button
            class="sign-in__btn"
            disabled=""
            type="submit"
          >
            Sign in
          </button>
        </div>
      </form>
    </div>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--light logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;

exports[`SignIn component Should render according to the snapshot while authorization status is unset or rejected 1`] = `
<div>
  <div
    class="user-page"
  >
    <header
      class="page-header user-page__head"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <h1
        class="page-title user-page__title"
      >
        Sign in
      </h1>
    </header>
    <div
      class="sign-in user-page__content"
    >
      <form
        action="#"
        class="sign-in__form"
      >
        <div
          class="sign-in__fields"
        >
          <div
            class="sign-in__field"
          >
            <input
              autocomplete="username"
              class="sign-in__input"
              id="user-email"
              name="user-email"
              placeholder="Email address"
              required=""
              type="email"
              value=""
            />
            <label
              class="sign-in__label visually-hidden"
              for="user-email"
            >
              Email address
            </label>
          </div>
          <div
            class="sign-in__field"
          >
            <input
              autocomplete="current-password"
              class="sign-in__input"
              id="user-password"
              name="user-password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
            <label
              class="sign-in__label visually-hidden"
              for="user-password"
            >
              Password
            </label>
          </div>
        </div>
        <div
          class="sign-in__submit"
        >
          <button
            class="sign-in__btn"
            type="submit"
          >
            Sign in
          </button>
        </div>
      </form>
    </div>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--light logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;

exports[`SignIn component Should render according to the snapshot while authorization status is unset or rejected 2`] = `
<div>
  <div
    class="user-page"
  >
    <header
      class="page-header user-page__head"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <h1
        class="page-title user-page__title"
      >
        Sign in
      </h1>
    </header>
    <div
      class="sign-in user-page__content"
    >
      <form
        action="#"
        class="sign-in__form"
      >
        <div
          class="sign-in__fields"
        >
          <div
            class="sign-in__field"
          >
            <input
              autocomplete="username"
              class="sign-in__input"
              id="user-email"
              name="user-email"
              placeholder="Email address"
              required=""
              type="email"
              value=""
            />
            <label
              class="sign-in__label visually-hidden"
              for="user-email"
            >
              Email address
            </label>
          </div>
          <div
            class="sign-in__field"
          >
            <input
              autocomplete="current-password"
              class="sign-in__input"
              id="user-password"
              name="user-password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
            <label
              class="sign-in__label visually-hidden"
              for="user-password"
            >
              Password
            </label>
          </div>
        </div>
        <div
          class="sign-in__submit"
        >
          <button
            class="sign-in__btn"
            type="submit"
          >
            Sign in
          </button>
        </div>
      </form>
    </div>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--light logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;

exports[`SignIn component Should render according to the snapshot while authorization status is unset or rejected 3`] = `
<div>
  <div
    class="user-page"
  >
    <header
      class="page-header user-page__head"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <h1
        class="page-title user-page__title"
      >
        Sign in
      </h1>
    </header>
    <div
      class="sign-in user-page__content"
    >
      <form
        action="#"
        class="sign-in__form"
      >
        <div
          class="sign-in__fields"
        >
          <div
            class="sign-in__field"
          >
            <input
              autocomplete="username"
              class="sign-in__input"
              id="user-email"
              name="user-email"
              placeholder="Email address"
              required=""
              type="email"
              value=""
            />
            <label
              class="sign-in__label visually-hidden"
              for="user-email"
            >
              Email address
            </label>
          </div>
          <div
            class="sign-in__field"
          >
            <input
              autocomplete="current-password"
              class="sign-in__input"
              id="user-password"
              name="user-password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
            <label
              class="sign-in__label visually-hidden"
              for="user-password"
            >
              Password
            </label>
          </div>
        </div>
        <div
          class="sign-in__submit"
        >
          <button
            class="sign-in__btn"
            type="submit"
          >
            Sign in
          </button>
        </div>
      </form>
    </div>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          aria-current="page"
          class="logo__link logo__link--light logo__link--active"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;
